### PR TITLE
MERGE ADD 2D PONG GAME:

### PR DESCRIPTION
**Pull request by nicolkaver**
Hello!
I am back with my 2d game with full history, I have managed to rebase master to my original branch (I had created a new branch last time missing the entire commit history last time) as well as to correct many eslint errors. I was not able to correct them all, mostly because in some cases I was not sure about what to do. Thus, do not hesitate to leave me a comment if you have the solution :)

You can now launch the game at localhost:3001/the-game. Following Alex's recommendations, its router is not anymore in VisitorRouter file but rather in LoggedinRouter. This is why should you try to launch it, follow the instructions here: 
https://github.com/AdrianWeaver/transcendence/pull/34

### CODE REVIEW

File: front/app/src/Views/GameCanvas/GameCanvas.tsx [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	const clear = () =>
+	{
+		if (game.board.ctx)
+		{
+			game.board.ctx.fillStyle = "#fff";
+			game.board.ctx?.clearRect(0, 0,
+				game.board.dim.width, game.board.dim.height);
+		}
+	};
+
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/GameCanvas.tsx:41]: Here the type of the variable is KeyboardEvent

File: front/app/src/Views/GameCanvas/GameCanvas.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	// I don't know what type we need to put here
+	const	keyHookDown = (e) =>
+	{
+		switch (e.keyCode)
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/GameCanvas.tsx:45]: e.keycode is now deprecated, so we need to use e.code
The type of code from KeyboardEvent is string, so need to make some adjustment in switch cases
"KeyS"
"KeyW"
"ArrowDown"
"ArrowUp"
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent

File: front/app/src/Views/GameCanvas/GameCanvas.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
+/* eslint-disable max-lines-per-function */
+/* eslint-disable max-statements */
+import { useEffect, useRef } from "react";
+
+// if we don't import all objects, the game will not work
+import Game from "./objects/Game";
+import Board from "./objects/Board";
```
[the remark this answers is not in the document]
Comment (nicolkaver) [reply front/app/src/Views/GameCanvas/GameCanvas.tsx]: Done, thank you !

File: front/app/src/Views/GameCanvas/GameCanvas.tsx (hunk 1 above)
Comment (nicolkaver) [reply front/app/src/Views/GameCanvas/GameCanvas.tsx:33]: Done.

File: front/app/src/Views/GameCanvas/GameCanvas.tsx (hunk 2 above)
Comment (nicolkaver) [reply front/app/src/Views/GameCanvas/GameCanvas.tsx]: Done.

**Comment by nicolkaver** [comment]
Thank you, Alex, for your comments, I have made changes as suggested. 

Moreover, I have also used this time to make the game stop whenever a player scores 7 points.

File: front/app/src/Views/GameCanvas/objects/Ball.tsx [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+			{
+				const direction = Math.floor(Math.random() * 10);
+				if (direction % 2 === 0)
+					this.angle = this.degreesToRadians(180);
+				else
+					this.angle = this.radiansToDegrees(0);
+			}
+			this.firstSetDirection = 0;
+		};
+        this.update = () =>
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/objects/Ball.tsx:70]: mixed tab style

File: front/app/src/Views/GameCanvas/GameCanvas.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		const pauseButton = pauseButtonRef.current;
+		const resumeButton = resumeButtonRef.current;
+
+		const ctx = canvas?.getContext("2d");
+		game.board.canvas = canvas;
+		game.board.ctx = ctx;
+		game.board.init();
+		addEventListener("keydown", keyHookDown);
+		addEventListener("keyup", keyHookReleased);
+		addEventListener("keydown", keyEnter);
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/GameCanvas.tsx:83]: you register the same event twice "keydown", you should detect "Enter" into keyHookDown

File: front/app/src/Views/GameCanvas/GameCanvas.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+				{
+					game.continueAnimating = false;
+					game.displayEndMessage();
+					game.ball.init();
+				}
+			}
+			requestId = requestAnimationFrame(render);
+		};
+		requestId = requestAnimationFrame(render);
+		return (() =>
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/GameCanvas.tsx:126]: this is executed on unmounting the component, you use cancelAnimationFam, that's good !
you forgot to removeEventListener at unmount
https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener

File: front/app/src/Views/GameCanvas/GameCanvas.tsx (hunk 5 above)
Comment (nicolkaver) [reply front/app/src/Views/GameCanvas/GameCanvas.tsx:83]: The game did not work anymore after I tried this

File: front/app/src/Views/GameCanvas/GameCanvas.tsx (hunk 6 above)
Comment (nicolkaver) [reply front/app/src/Views/GameCanvas/GameCanvas.tsx:126]: Thank you for this advice, I did not know it even existed, I have added it now.

File: front/app/src/Views/GameCanvas/objects/Ball.tsx (hunk 4 above)
Comment (nicolkaver) [reply front/app/src/Views/GameCanvas/objects/Ball.tsx]: Done.

File: front/app/src/Views/GameCanvas/objects/Ball.tsx [hunk 7]
```diff
@@ -0,0 +1,205 @@
+/* eslint-disable curly */
+/* eslint-disable max-statements */
+/* eslint-disable max-lines-per-function */
+// import { ConstructionOutlined } from "@mui/icons-material";
+import Game from "./Game";
+import Position from "./Position";
+
+class Ball
+{
+    public pos: Position;
```
[the remark this answers is not in the document]
Comment (nicolkaver) [reply front/app/src/Views/GameCanvas/objects/Ball.tsx]: Done.

**Comment by nicolkaver** [comment]
Hello, I have made some minor changes to the ball speed. It isn't much, it is just to make it look more realistic. I tried to do some more developed stuff but did not succeed. If you have a better solution, do not hesitate to apply it.

File: front/app/src/Router/LoggedRouter.tsx [hunk 8]
```diff
@@ -61,6 +62,11 @@ const	LoggedRouter = () =>
 					>
 
 				</Route>
+				<Route
+					path="the-game"
+					element={<GameCanvas />}
+				/>
+
```
Comment (apayetRun) [inline front/app/src/Router/LoggedRouter.tsx:69]: We've got trailing space here, need just a small fix :)

File: front/app/src/Views/GameCanvas/objects/Ball.tsx [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+		};
+	}
+}
+
+export default Ball;
+
+
+/*
+computeSpeed
+*/
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/objects/Ball.tsx:207]: No newline at the end of this file

Comment (apayetRun) [reply front/app/src/Views/GameCanvas/objects/Ball.tsx:207]: Maybe comment are useless ?

File: front/app/src/Views/GameCanvas/objects/Board.tsx [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+			{
+				this.game.ball.pos.x *= multiplicatorWidth;
+				this.game.ball.pos.y *= multiplicatorHeight;
+			}
+		};
+		this.registerEvents = () =>
+		{
+			addEventListener("resize", this.updateSizeAfterResize);
+		};
+		// eslint-disable-next-line max-statements
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/objects/Board.tsx:92]: this one is no longer used, you can safely remove it.

File: front/app/src/Views/GameCanvas/objects/Dimension.tsx [hunk 11]
```diff
@@ -0,0 +1,19 @@
+class Dimension
+{
+    public height: number;
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/objects/Dimension.tsx:3]: We've got space instead tab

File: front/app/src/Views/GameCanvas/objects/Ball.tsx [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+					/ (this.game.playerTwo.racket.dim.height / 2);
+			let bounceAngle;
+			bounceAngle = 0;
+			if (normalizedRelativeIntersectionY)
+			{
+				bounceAngle = normalizedRelativeIntersectionY
+					* this.radiansToDegrees((Math.PI / 4));
+				if (normalizedRelativeIntersectionY < 0)
+					normalizedRelativeIntersectionY *= -1;
+				if (normalizedRelativeIntersectionY <= 0.15)
```
Comment (apayetRun) [inline front/app/src/Views/GameCanvas/objects/Ball.tsx:188]: This part must be fixed, maybe on a dedicated PR,
I think we have acceleration when hit the paddle, 
and decelleration should be when paddle angle reflected is the oposite of this angle of the ball incomming. (with a gradient that apply this effect)
Example Incoming angle (not real value just for explanation) : 180 and reflected angle 0 the ball go the the same direction it come.
This example value is not realistic, just to have the opposite value. 
Maybe we can define what is the angle from the paddle that triger this feature and a gradient to have a coef from the center of the paddle to make less or more effect.

**Comment by AdrianWeaver** [comment]
Closing pull request due to invalid branch.
Please see above comments and discord discussion.

**Pull request closed without merging**